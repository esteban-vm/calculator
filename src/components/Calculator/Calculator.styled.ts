import { styled, css, mediaQuery } from '@/styles'

export const Wrapper = styled.main`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
  background: linear-gradient(to right, var(--color-lighter-blue), var(--color-darker-blue));
`

export const Container = styled.section`
  display: grid;
  grid-template-columns: repeat(4, 25%);
  grid-template-rows: repeat(7, calc(100% / 7));
  width: 400px;
  height: 650px;
  box-shadow: 2px 2px 10px var(--color-darker);
  border-radius: 10px;

  ${mediaQuery('touch')} {
    width: 90%;
    height: 70%;

    @media (orientation: landscape) {
      width: 50%;
      height: 90%;
    }
  }
`

export const Screen = styled.header`
  grid-column: 1 / -1;
  grid-row: 1 / span 2;
  display: flex;
  justify-content: space-around;
  flex-direction: column;
  padding: 10px;
  background-color: var(--color-dark);
  word-wrap: break-word;
  word-break: break-all;
  text-align: right;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
`

export const Previous = styled.div`
  color: var(--color-light);
  font-size: 1.5rem;
`

export const Current = styled.div`
  color: var(--color-white);
  font-size: 3rem;
`

type ButtonProps = { spanned?: boolean; rounded?: 'right' | 'left'; operation?: boolean; control?: boolean }

export const Button = styled.button<ButtonProps>`
  cursor: pointer;
  user-select: none;
  font-size: 2rem;
  border: 1px outset var(--color-white);
  outline: none;

  :active {
    transform: scale(0.98);
  }

  ${mediaQuery('mouse')} {
    :hover {
      background-color: var(--color-lighter);
    }
  }

  ${({ spanned }) => {
    if (spanned) {
      return css`
        grid-column: span 2;
      `
    }
  }}

  ${({ rounded }) => {
    if (rounded) {
      return css`
        border-bottom-${rounded}-radius: 10px;
      `
    }
  }}

  ${({ control, operation }) => {
    return css`
      background-color: ${control ? 'skyblue' : operation ? 'gray' : 'var(--color-light)'};
    `
  }}
`
