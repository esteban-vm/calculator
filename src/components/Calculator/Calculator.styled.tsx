import type { FC, MouseEventHandler, HTMLAttributes } from 'react'
import { styled, css } from '@/styles'

export const Wrapper = styled.main`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100vw;
  height: 100vh;

  ${({ theme }) => {
    return css`
      background: linear-gradient(to right, ${theme.colors.color1}, ${theme.colors.color2});
    `
  }}
`

export const Container = styled.section`
  display: grid;
  grid-template-columns: repeat(4, 25%);
  grid-template-rows: repeat(7, calc(100% / 7));
  box-shadow: 2px 2px 10px #333;
  border-radius: 10px;
  width: 400px;
  height: 650px;

  @media (pointer: coarse) {
    @media (orientation: portrait) {
      width: 90%;
      height: 70%;
    }

    @media (orientation: landscape) {
      width: 50%;
      height: 90%;
    }
  }
`

export const Screen = styled.header`
  grid-column: 1 / -1;
  grid-row: 1 / span 2;
  display: flex;
  justify-content: space-around;
  flex-direction: column;
  padding: 10px;
  background-color: #3c4043bf;
  word-wrap: break-word;
  word-break: break-all;
  text-align: right;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
`

export const Previous = styled.div`
  color: #ffffffbf;
  font-size: 1.5rem;
`

export const Current = styled.div`
  color: #fff;
  font-size: 3rem;
`

interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
  onClick: MouseEventHandler<HTMLButtonElement>
  'data-key': string
  title: string
  rounded?: 'right' | 'left'
  operation?: boolean
  control?: boolean
}

export const Button: FC<ButtonProps> = ({ onClick, ...rest }) => {
  const handleClick: typeof onClick = (event) => {
    onClick(event)
    navigator.vibrate?.(50)
  }

  return <StyledButton onClick={handleClick} {...rest} />
}

const StyledButton = styled.button<ButtonProps>`
  cursor: pointer;
  user-select: none;
  font-size: 2rem;
  border: 1px outset #fff;
  outline: none;

  @media (pointer: fine) {
    :hover {
      background-color: #ffffffe6;
    }
  }

  ${({ rounded }) => {
    if (rounded) {
      return css`
        border-bottom-${rounded}-radius: 10px;
      `
    }
  }}

  ${({ control, operation, theme }) => {
    return css`
      background-color: ${theme.colors[control ? 'color4' : operation ? 'color3' : 'color5']};
    `
  }}
`
