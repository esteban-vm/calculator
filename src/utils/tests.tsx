/* eslint-disable react-refresh/only-export-components */
import type { FC, PropsWithChildren, ReactElement } from 'react'
import { render, type RenderOptions } from '@testing-library/react'
import { ThemeProvider, themes } from '@/styles'

const Wrapper: FC<PropsWithChildren> = ({ children }) => {
  return <ThemeProvider theme={themes['light']}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions) => {
  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render }
export { default as userEvent } from '@testing-library/user-event'
