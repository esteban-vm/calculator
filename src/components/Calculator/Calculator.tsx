import type { FC } from 'react'
import { useCalculator, useTheme } from '@/hooks'
import * as Styled from './Calculator.styled'

const Calculator: FC = () => {
  const { toggle: toggleTheme } = useTheme()
  const { previous, current, operation, appendValue, clearEntry, clearAll, chooseOperation, getResult, toggleSign } =
    useCalculator()

  return (
    <Styled.Wrapper>
      <Styled.Container aria-label='Calculator'>
        <Styled.Screen>
          <Styled.Previous>
            {previous} {operation}
          </Styled.Previous>
          <Styled.Current>{current}</Styled.Current>
        </Styled.Screen>
        <Styled.Button onClick={clearAll} data-key='shift+backspace' title='All clear (shift+backspace)' control>
          AC
        </Styled.Button>
        <Styled.Button onClick={clearEntry} data-key='backspace' title='Clear entry (backspace)' control>
          CE
        </Styled.Button>
        <Styled.Button onClick={toggleSign} data-key='shift+s' title='Toggle sign (shift+s)' control>
          +/-
        </Styled.Button>
        <Styled.Button onClick={chooseOperation} data-key='/' title='Divide' operation>
          รท
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='7' title='Seven'>
          7
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='8' title='Eight'>
          8
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='9' title='Nine'>
          9
        </Styled.Button>
        <Styled.Button onClick={chooseOperation} data-key='*' title='Multiply' operation>
          ร
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='4' title='Four'>
          4
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='5' title='Five'>
          5
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='6' title='Six'>
          6
        </Styled.Button>
        <Styled.Button onClick={chooseOperation} data-key='plus' title='Add' operation>
          +
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='1' title='One'>
          1
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='2' title='Two'>
          2
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='3' title='Three'>
          3
        </Styled.Button>
        <Styled.Button onClick={chooseOperation} data-key='minus' title='Subtract' operation>
          -
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='.' title='Decimal point' rounded='left' control>
          .
        </Styled.Button>
        <Styled.Button onClick={appendValue} data-key='0' title='Zero'>
          0
        </Styled.Button>
        <Styled.Button onClick={toggleTheme} data-key='shift+t' title='Toggle theme (shift+t)' control>
          <svg viewBox='0 0 24 24' fill='currentColor' width='3rem' height='3rem'>
            <path d='M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 017.5 2m11.57 1.5l1.43 1.43L4.93 20.5 3.5 19.07 19.07 3.5m-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13.51 1.67m-3.3 3.61l-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87.4 1.31M19 13.5a5.5 5.5 0 01-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58l2.77-1.15-.24 3.35-2.53-2.2m4.33-2.7l1.15-2.77 2.2 2.54-3.35.23m1.15-4.96l-1.14-2.78 3.34.24-2.2 2.54M9.63 18.93l2.77 1.15-2.53 2.19-.24-3.34z' />
          </svg>
        </Styled.Button>
        <Styled.Button onClick={getResult} data-key='enter' title='Equals' rounded='right' operation>
          =
        </Styled.Button>
      </Styled.Container>
    </Styled.Wrapper>
  )
}

export default Calculator
