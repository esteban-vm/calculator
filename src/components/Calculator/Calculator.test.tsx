import { render, cleanup, screen, userEvent } from '@/tests'
import Calculator from './Calculator'

describe('CALCULATOR TEST SUITE:', () => {
  let asFragment: () => DocumentFragment

  beforeEach(() => {
    void ({ asFragment } = render(<Calculator />))
  })

  afterEach(cleanup)

  describe('style tests:', () => {
    it('should have correct styles', () => {
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('render tests:', () => {
    it('should render the buttons for numbers', () => {
      const numbers = [...Array(10).keys()]
      for (const num of numbers) expect(screen.getByRole('button', { name: String(num) })).toBeInTheDocument()
    })

    it('should render the button for all clear', () => {
      expect(screen.getByRole('button', { name: /ac/i, description: /all clear/i })).toBeInTheDocument()
    })

    it('should render the button to clear entry', () => {
      expect(screen.getByRole('button', { name: /ce/i, description: /clear entry/i })).toBeInTheDocument()
    })

    it('should render the button to toggle sign', () => {
      expect(screen.getByRole('button', { name: /\+\/-/, description: /toggle sign/i })).toBeInTheDocument()
    })

    it('should render the button for division', () => {
      expect(screen.getByRole('button', { name: /÷/, description: /divide/i })).toBeInTheDocument()
    })

    it('should render the button for multiplication', () => {
      expect(screen.getByRole('button', { name: /×/, description: /multiply/i })).toBeInTheDocument()
    })

    it('should render the button for addition', () => {
      expect(screen.getByRole('button', { name: /\+/, description: /add/i })).toBeInTheDocument()
    })

    it('should render the button for subtraction', () => {
      expect(screen.getByRole('button', { name: /-/, description: /subtract/i })).toBeInTheDocument()
    })

    it('should render the button for decimal point', () => {
      expect(screen.getByRole('button', { name: /\./, description: /decimal point/i })).toBeInTheDocument()
    })

    it('should render the button to toggle theme', () => {
      expect(screen.getByRole('button', { description: /toggle theme/i })).toBeInTheDocument()
    })

    it('should render the button for equals', () => {
      expect(screen.getByRole('button', { name: /=/, description: /equals/i })).toBeInTheDocument()
    })
  })

  describe('control tests:', () => {
    it('should vibrate when pressing a button', async () => {
      await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      expect(navigator.vibrate).toHaveBeenCalledWith(50)
    })

    it('should not be able to write numbers of more than 10 digits', async () => {
      for (let time = 1; time <= 11; time++) {
        await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      }
      expect(await screen.findByText(/1{10}/)).toBeInTheDocument()
    })

    it('should be able to clear all (e.g. 51 => "")', async () => {
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      expect(await screen.findByText(/51/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /ac/i, description: /all clear/i }))
      expect(screen.queryByText(/51/)).not.toBeInTheDocument()
    })

    it('should be able to clear entry (e.g. 958 => 95)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /9/, description: /nine/i }))
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      await userEvent.click(screen.getByRole('button', { name: /8/, description: /eight/i }))
      expect(await screen.findByText(/958/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /ce/i, description: /clear entry/i }))
      expect(screen.queryByText(/958/)).not.toBeInTheDocument()
      expect(await screen.findByText(/95/)).toBeInTheDocument()
    })

    it('should be able to toggle sign (e.g. 80 => -80)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /8/, description: /eight/i }))
      await userEvent.click(screen.getByRole('button', { name: /0/, description: /zero/i }))
      await userEvent.click(screen.getByRole('button', { name: /\+\/-/, description: /toggle sign/i }))
      expect(await screen.findByText(/-80/)).toBeInTheDocument()
    })

    it('should not be able to toggle sign if current value does not exist', async () => {
      await userEvent.click(screen.getByRole('button', { name: /\+\/-/, description: /toggle sign/i }))
      expect((await screen.findByRole('banner')).children[1]).not.toHaveTextContent('-')
    })

    it('should not be able to choose operation if current value does not exist', async () => {
      await userEvent.click(screen.getByRole('button', { name: /\+/, description: /add/i }))
      expect((await screen.findByRole('banner')).children[0]).not.toHaveTextContent('+')
    })

    it('should not be able to write a zero as first digit (e.g. 050 => 50)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /0/, description: /zero/i }))
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      await userEvent.click(screen.getByRole('button', { name: /0/, description: /zero/i }))
      expect(screen.queryByText(/050/)).not.toBeInTheDocument()
      expect(await screen.findByText(/50/)).toBeInTheDocument()
    })

    it('should be able to append a decimal point (e.g. 5.9)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      await userEvent.click(screen.getByRole('button', { name: /\./, description: /decimal point/i }))
      await userEvent.click(screen.getByRole('button', { name: /9/, description: /nine/i }))
      expect(await screen.findByText(/5\.9/)).toBeInTheDocument()
    })

    it('should be able to append a decimal point preceded by a zero (e.g. "." => 0.)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /\./, description: /decimal point/i }))
      expect(await screen.findByText(/0\./)).toBeInTheDocument()
    })

    it('should not be able to append a decimal point if already there is one', async () => {
      await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      await userEvent.click(screen.getByRole('button', { name: /\./, description: /decimal point/i }))
      await userEvent.click(screen.getByRole('button', { name: /\./, description: /decimal point/i }))
      expect(screen.queryByText(/1\.\./)).not.toBeInTheDocument()
      expect(await screen.findByText(/1\./)).toBeInTheDocument()
    })

    it('should be able to toggle theme', async () => {
      expect(localStorage.getItem('calculator:theme')).toBe('"light"')
      await userEvent.click(screen.getByRole('button', { description: /toggle theme/i }))
      expect(localStorage.getItem('calculator:theme')).toBe('"dark"')
    })
  })

  describe('operation tests:', () => {
    it('should add two numbers (e.g. 53 + 17)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      await userEvent.click(screen.getByRole('button', { name: /3/, description: /three/i }))
      expect(await screen.findByText(/53/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /\+/, description: /add/i }))
      await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      await userEvent.click(screen.getByRole('button', { name: /7/, description: /seven/i }))
      expect(await screen.findByText(/17/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /=/, description: /equals/i }))
      expect(await screen.findByText(/70/)).toBeInTheDocument()
    })

    it('should subtract two numbers (e.g. 75 - 10)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /7/, description: /seven/i }))
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      expect(await screen.findByText(/75/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /-/, description: /subtract/i }))
      await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      await userEvent.click(screen.getByRole('button', { name: /0/, description: /zero/i }))
      expect(await screen.findByText(/10/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /=/, description: /equals/i }))
      expect(await screen.findByText(/65/)).toBeInTheDocument()
    })

    it('should multiply two numbers (e.g. 20 × 15)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /2/, description: /two/i }))
      await userEvent.click(screen.getByRole('button', { name: /0/, description: /zero/i }))
      expect(await screen.findByText(/20/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /×/, description: /multiply/i }))
      await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      expect(await screen.findByText(/15/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /=/, description: /equals/i }))
      expect(await screen.findByText(/300/)).toBeInTheDocument()
    })

    it('should divide two numbers (e.g. 325 ÷ 25)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /3/, description: /three/i }))
      await userEvent.click(screen.getByRole('button', { name: /2/, description: /two/i }))
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      expect(await screen.findByText(/325/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /÷/, description: /divide/i }))
      await userEvent.click(screen.getByRole('button', { name: /2/, description: /two/i }))
      await userEvent.click(screen.getByRole('button', { name: /5/, description: /five/i }))
      expect(await screen.findByText(/^25/)).toBeInTheDocument()
      await userEvent.click(screen.getByRole('button', { name: /=/, description: /equals/i }))
      expect(await screen.findByText(/13/)).toBeInTheDocument()
    })

    it('should not divide by zero', async () => {
      await userEvent.click(screen.getByRole('button', { name: /1/, description: /one/i }))
      await userEvent.click(screen.getByRole('button', { name: /÷/, description: /divide/i }))
      await userEvent.click(screen.getByRole('button', { name: /0/, description: /zero/i }))
      await userEvent.click(screen.getByRole('button', { name: /=/, description: /equals/i }))
      expect(screen.queryByText(Infinity.toString())).not.toBeInTheDocument()
    })

    it('should be able to perform all operations (e.g. (6 × 4 - 3 + 8) ÷ 9)', async () => {
      await userEvent.click(screen.getByRole('button', { name: /6/, description: /six/i }))
      await userEvent.click(screen.getByRole('button', { name: /×/, description: /multiply/i }))
      await userEvent.click(screen.getByRole('button', { name: /4/, description: /four/i }))
      await userEvent.click(screen.getByRole('button', { name: /-/, description: /subtract/i }))
      await userEvent.click(screen.getByRole('button', { name: /3/, description: /three/i }))
      await userEvent.click(screen.getByRole('button', { name: /\+/, description: /add/i }))
      await userEvent.click(screen.getByRole('button', { name: /8/, description: /eight/i }))
      await userEvent.click(screen.getByRole('button', { name: /÷/, description: /divide/i }))
      await userEvent.click(screen.getByRole('button', { name: /9/, description: /nine/i }))
      await userEvent.click(screen.getByRole('button', { name: /=/, description: /equals/i }))
      expect(await screen.findByText(/3\.22/)).toBeInTheDocument()
    })
  })
})
