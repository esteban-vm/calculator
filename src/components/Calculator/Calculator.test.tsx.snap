// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CALCULATOR TEST SUITE: > style tests: > should have correct styles 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
  background: linear-gradient(to right, #ff3f7b, #ff766e);
}

.emotion-1 {
  display: grid;
  grid-template-columns: repeat(4, 25%);
  grid-template-rows: repeat(7, calc(100% / 7));
  width: 400px;
  height: 650px;
  box-shadow: 2px 2px 10px #333;
  border-radius: 10px;
}

@media (hover: none), (pointer: coarse) {
  .emotion-1 {
    width: 90%;
    height: 70%;
  }

  @media (orientation: landscape) {
    .emotion-1 {
      width: 50%;
      height: 90%;
    }
  }
}

.emotion-2 {
  grid-column: 1/-1;
  grid-row: 1/span 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -ms-flex-pack: space-around;
  -webkit-justify-content: space-around;
  justify-content: space-around;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 10px;
  background-color: #3c4043bf;
  word-wrap: break-word;
  word-break: break-all;
  text-align: right;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
}

.emotion-3 {
  color: #ffffffbf;
  font-size: 1.5rem;
}

.emotion-4 {
  color: #fff;
  font-size: 3rem;
}

.emotion-5 {
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 2rem;
  border: 1px outset #fff;
  outline: none;
  background-color: #fffbc4;
}

.emotion-5:active {
  -webkit-transform: scale(0.98);
  -moz-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
}

@media (hover: hover) and (pointer: fine) {
  .emotion-5:hover {
    background-color: #ffffffe6;
  }
}

.emotion-8 {
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 2rem;
  border: 1px outset #fff;
  outline: none;
  background-color: #ffc78c;
}

.emotion-8:active {
  -webkit-transform: scale(0.98);
  -moz-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
}

@media (hover: hover) and (pointer: fine) {
  .emotion-8:hover {
    background-color: #ffffffe6;
  }
}

.emotion-9 {
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 2rem;
  border: 1px outset #fff;
  outline: none;
  background-color: #fffbe7;
}

.emotion-9:active {
  -webkit-transform: scale(0.98);
  -moz-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
}

@media (hover: hover) and (pointer: fine) {
  .emotion-9:hover {
    background-color: #ffffffe6;
  }
}

.emotion-21 {
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 2rem;
  border: 1px outset #fff;
  outline: none;
  border-bottom-left-radius: 10px;
  background-color: #fffbc4;
}

.emotion-21:active {
  -webkit-transform: scale(0.98);
  -moz-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
}

@media (hover: hover) and (pointer: fine) {
  .emotion-21:hover {
    background-color: #ffffffe6;
  }
}

.emotion-24 {
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 2rem;
  border: 1px outset #fff;
  outline: none;
  border-bottom-right-radius: 10px;
  background-color: #ffc78c;
}

.emotion-24:active {
  -webkit-transform: scale(0.98);
  -moz-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
}

@media (hover: hover) and (pointer: fine) {
  .emotion-24:hover {
    background-color: #ffffffe6;
  }
}

<main
    class="emotion-0"
  >
    <section
      aria-label="Calculator"
      class="emotion-1"
    >
      <header
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
           
        </div>
        <div
          class="emotion-4"
        />
      </header>
      <button
        class="emotion-5"
        title="All clear"
      >
        AC
      </button>
      <button
        class="emotion-5"
        title="Clear entry"
      >
        CE
      </button>
      <button
        class="emotion-5"
        title="Toggle sign"
      >
        +/-
      </button>
      <button
        class="emotion-8"
        title="Divide"
      >
        Ã·
      </button>
      <button
        class="emotion-9"
      >
        7
      </button>
      <button
        class="emotion-9"
      >
        8
      </button>
      <button
        class="emotion-9"
      >
        9
      </button>
      <button
        class="emotion-8"
        title="Multiply"
      >
        Ã—
      </button>
      <button
        class="emotion-9"
      >
        4
      </button>
      <button
        class="emotion-9"
      >
        5
      </button>
      <button
        class="emotion-9"
      >
        6
      </button>
      <button
        class="emotion-8"
        title="Add"
      >
        +
      </button>
      <button
        class="emotion-9"
      >
        1
      </button>
      <button
        class="emotion-9"
      >
        2
      </button>
      <button
        class="emotion-9"
      >
        3
      </button>
      <button
        class="emotion-8"
        title="Subtract"
      >
        -
      </button>
      <button
        class="emotion-21"
        title="Decimal point"
      >
        .
      </button>
      <button
        class="emotion-9"
      >
        0
      </button>
      <button
        class="emotion-5"
        title="Toggle theme"
      >
        ðŸŒ‘
      </button>
      <button
        class="emotion-24"
        title="Equals"
      >
        =
      </button>
    </section>
  </main>
</DocumentFragment>
`;
